mautic/mautic
# Campaign duplicate entry errors

closes https://github.com/mautic/mautic/issues/6613
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | N
| Automated tests included? | N
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
When running `console mautic:campaign:trigger` there are a few situations that can result in an integrity constraint violation AKA "Duplicate entry" errors. These *appear* innocuous and are generally ignored by people, but in some cases they can cause serious problems with lead handling. Here are some examples:

1. **User Error:** You are running multiple concurrent instances of the cron task `console mautic:campaign:trigger` probably also using --force, but without using the multi-threading parameters correctly. These processes are hitting the same event, lead and rotation at the same time. This PR will detect this as two events matching each other within a 60s timeframe and prevent saving a new event, thus preventing the exception most of the time (though not always). Note: Use --thread-id and --max-threads to avoid this problem altogether.
2. **Jump To Events:** You have jump-to event that takes leads somewhere they've already been, by way of intervals (particularly if there are intervals on both sides). This will cause the lead to stop when jumping to the event they've already hit. This PR fixes this by increasing the rotation before the jump occurs instead of afterward. This PR may resolve #7129.
3. **Event Rearrangement:** You have a campaign with a delay in a chain of events, and you re-arrange the order of events while leads are in the campaign queued for an event. It's like playing with a bug, and you have shuffled your hands to keep the bug on them as it walks across. In this case, the bug is the lead and your hands are two events. Previously the lead would HALT at the first "duplicate" event, and not run further through that branch of the campaign. This is not expected behavior by users. We have some events with hundreds of thousands of leads queued to continue after a delay, and the campaign manager (cognizant of this) may intentionally rearrange events to have the lead go back to a previous event in order to handle campaign changes gracefully. This PR fixes this by detecting this situation and incrementing the rotation when it occurs on a per event-lead-rotation basis.

Finally, this PR detects these critical DB related errors and throws them before they have a chance to mess up subsequent batches of leads. A single integrity constraint violation in the midst of a large batch of leads in a complex campaign can cause duplicate events (such as push to integration) to be fired in the thousands, so throwing the exception so that a new cron/thread picks up the body of work is much safer for us.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
* Run multiple simultaneous threads of `console mautic:campaign:trigger --force`. OR
* Rearrange a campaign vertically while leads are in flight and there are delays between events. OR
* Create a delayed "Jump To" event that sends a lead to an event that they have already hit. OR

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Make a simple campaign with 2 actions chained together, and a delay on the second action of a few minutes.
3. Add a lead to the campaign, wait till the lead is scheduled for the second event. (crons run every 5 minutes)
4. Edit the campaign (while the lead is in the middle), and change the order of events (first is second, and vice versa).
5. Wait ~5 minutes.
6. Check the lead, it should have fired one of those events TWICE. This is the expected behavior.
